### PR TITLE
[bbc] Extract more metadata for window.__INITIAL_DATA__ pages

### PR DESCRIPTION
**Pull request by dirkf**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Pages with metadata extracted from a JSON object assigned to window.__INITIAL_DATA__ didn't extract timestamp and description. At least for the type where the metadata is in the `media-experience` object, this PR adds those items.

Examples:
* http://www.bbc.co.uk/news/science-environment-33661876 (existing test URL)
* http://www.bbc.com/news/av/10462520 (issue #28746)

### CODE REVIEW

File: youtube_dl/extractor/bbc.py [hunk 1]
```diff
@@ -1164,12 +1170,20 @@ def parse_media(media):
...
+                            item_time = parse_iso8601(meta.get('timestamp'))
+                            break
                     entries.append({
                         'id': item_id,
                         'title': item_title,
                         'thumbnail': item.get('holdingImageUrl'),
                         'formats': formats,
                         'subtitles': subtitles,
+                        'timestamp': item_time,
+                        'description': item_desc.strip() if item_desc else None,
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py:1186]: `strip_or_none`.

File: youtube_dl/extractor/bbc.py [hunk 2]
```diff
@@ -1164,12 +1170,20 @@ def parse_media(media):
                         continue
                     formats, subtitles = self._download_media_selector(item_id)
                     self._sort_formats(formats)
+                    item_desc = try_get(media, lambda x: x['summary']['blocks'][0]['model']['text'], compat_str)
+                    item_time = None
+                    for meta in try_get(media, lambda x: x['metadata']['items'], list) or []:
+                        if try_get(meta, lambda x: x['label']) == 'Published':
+                            item_time = parse_iso8601(meta.get('timestamp'))
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py:1177]: `unified_timestamp`.

File: youtube_dl/extractor/bbc.py [hunk 3]
```diff
@@ -761,8 +762,13 @@ class BBCIE(BBCCoUkIE):
         'only_matching': True,
     }, {
         # custom redirection to www.bbc.com
+        # video with window.__INITIAL_DATA__
         'url': 'http://www.bbc.co.uk/news/science-environment-33661876',
-        'only_matching': True,
+        'id': 'p02xzws1',
+        'title': 'Pluto may have \'nitrogen glaciers\'',
+        'description': 'Pluto could have glaciers of nitrogen ice, new photographs from Nasa\'s New Horizons probe suggest.',
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py:769]: Use outer double quotes instead of escaping single inner quotes.

File: youtube_dl/extractor/bbc.py [hunk 4]
```diff
@@ -761,8 +762,13 @@ class BBCIE(BBCCoUkIE):
...
+        'thumbnail': 'https://c.files.bbci.co.uk/29DE/production/_84481701_84481700.jpg',
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py:770]: Do not use exact URLs in tests.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/bbc.py (hunk 1 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py:1186]: Sure, didn't know about that one.

File: youtube_dl/extractor/bbc.py (hunk 2 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py:1177]: Fine, a more general solution that I actually had originally but changed back, `parse_iso8601()` being already imported.

File: youtube_dl/extractor/bbc.py (hunk 4 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py:770]: How about a regex there, say `r're:https?://.+/.+\.jpg'`?

Thx for review, I'll push an update and post you when it's up.

Comment (dstftw) [reply youtube_dl/extractor/bbc.py:770]: >How about a regex there, say `r're:https?://.+/.+\.jpg'`?

This is fine.

File: youtube_dl/extractor/bbc.py (hunk 3 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py]: The escaped quotes were in the program output, but of course that's JSON and not Python ...

Comment (dstftw) [reply youtube_dl/extractor/bbc.py]: The point is to always have copy-paste ready string literal in tests for debugging/testing purposes.

File: youtube_dl/extractor/bbc.py (hunk 4 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py]: Seems to be a recurrent CI problem where `pip install nose` fails in the Jython VMs? If that's going to work next time, I'd like to push a rebased update.

File: youtube_dl/extractor/bbc.py [hunk 5]
```diff
@@ -761,8 +764,13 @@ class BBCIE(BBCCoUkIE):
...
     }, {
         # custom redirection to www.bbc.com
+        # video with window.__INITIAL_DATA__
         'url': 'http://www.bbc.co.uk/news/science-environment-33661876',
-        'only_matching': True,
+        'id': 'p02xzws1',
+        'title': "Pluto may have 'nitrogen glaciers'",
+        'description': "Pluto could have glaciers of nitrogen ice, new photographs from Nasa's New Horizons probe suggest.",
+        'thumbnail': r're:https?://.+/.+\.jpg',
+        'timestamp': 1437785037,
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py:773]: This must be inside a dict corresponding to `info_dict` key. `python -W error .\test\test_download.py TestDownload.test_BBC_13` must succeed.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/bbc.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/bbc.py]: Jython issue is intermittent you can ignore it. Also there is no need to rebase as GitHub does this just fine.

File: youtube_dl/extractor/bbc.py (hunk 5 above)
Comment (dirkf) [reply youtube_dl/extractor/bbc.py]: Obvs, stupid cut+paste error, thanks.

**Comment by dirkf** [comment]
Weirdly the changes were overwritten, but now restored,

**Comment by dirkf** [comment]
Thanks for your work, Sergey

**Pull request merged**